include/graaflib/algorithm/topological_sorting.tpp

```diff
@@ -0,0 +1,60 @@
+#pragma once
+
+#include <graaflib/algorithm/cycle_detection.h>
+#include <graaflib/algorithm/topological_sorting.h>
+
+#include <vector>
+#include <unordered_map>
+
+namespace graaf::algorithm {
+
+namespace detail {
+
+enum class vertex_status { UNVISITED, VISITED};
+
+// DFS topological sort
+template <typename V, typename E>
+void do_dfs_topological_sort(
+    const graph<V, E, graph_type::DIRECTED>& graph,
+    vertex_id_t start_vertex,
+    std::unordered_map<vertex_id_t, vertex_status>& processed_vertices,
+    std::vector<vertex_id_t>& sorted_vertices) {
+    
+    processed_vertices[start_vertex] = vertex_status::VISITED;
+    for (const auto& next_vertex : graph.get_neighbors(start_vertex)) {
+        if (processed_vertices[next_vertex] == vertex_status::UNVISITED) {
+            do_dfs_topological_sort(graph, next_vertex, processed_vertices,
+                sorted_vertices);
+        }
+    }
+
+        sorted_vertices.push_back(start_vertex);
+    }
+
+}; // namespace detail
+
+template <typename V, typename E>
+std::vector<vertex_id_t> topological_sort(
+    const graph<V, E, graph_type::DIRECTED>& graph) {
+    
+    // Graph should be acyclic
+    if (dfs_cycle_detection(graph)) {
+        return {};    
```

Comment:
Minor food for thought: Maybe there is a possibility to check for cycles during the topological sorting itself, which then might have better performance than the two-step approach. We can think about checking this in a follow up PR

Reply:
Agreed, in terms of O(n), it's still n, but in the case of a large number of vertices, it will affect performance.

Reply:
Good idea! I would indeed be fine with doing this in a follow-up, as the current PR already adds value 👍🏻

Reply:
I was thinking about how to combine cycle detection and node pushback traversal without using enums, etc. What do you think about using unordered_set as the return type from the function instead of vector? So basically, in the case of a cycle, vertex should be visited but wasn't present in the set for return type.